### PR TITLE
 kamailio-kemi-framework: All modules documented

### PR DESCRIPTION
**Pull request by verticelo**
Documentation for exported KEMI functions was added for all modules except the Core/HDR submodules as they are already added. 

If this is a route we want to go I will add additional functionality so we can customize the documentation for each function in the modules list as well. I have a few ideas on how to improve it further and make it more usable but this PR proposes the basic structure.

I am yet to commit the updated version of the python script, need to clean it up a little bit but wanted to present the documentation at least.

I'd like to receive comments on the structure of the documentation (markdown) itself, if the format is good or if it should be changed?

### CODE REVIEW

**Comment by miconda** [comment]
This looks very useful, thanks!

My opinions about the output format and structure:

  * I would prefer one file per module (e.g., acc.md, auth.md) instead of all functions listed in the same file (one big modules.md) -- the reasons behind are mainly the fact that updating is easier to add to a smaller file, but also easier to track changes. We have the same for docs of the modules for native cfg
  * I would still keep the prototype of kemi function, because sometimes is not one-to-one relation with the native cfg function. Also, in native cfg, functions were supposed to have only string parameters, so even now most of them are like this. With kemi I tried to use directly int when it was the case.

**Comment by verticelo** [comment]
Hey, I hope its OK to have a differing opinion and some counter-questions? :)

Here goes:
* I would still keep the prototype of kemi function..
-- Actually, everything in modules.md right now is from the prototypes. I believe every parameter to be 100% accurate towards the function called upon in the code. Have you seen anything that is contrary to this? If so, please give an example.

* I would prefer one file per module...
-- I had a few thoughts of why I thought the one file was nice:
-- When entering the tutorial page, only having ~10 menu options makes it a lot more digestible and readable than having 50+ menu options.
-- Auto-generation makes it easier to have it in one file rather than keeping track of which files should be added/removed, menu file updated etc
-- It is easier to use the function export documentation as a reference page. One can easily search for a function name and read about it rather than clicking around to different modules and wait for page loads etc.

I thought of adding to the script so for each module and function we look for the following files:
/kamailio-kemi-framework/docs/modules/modulename.md
/kamailio-kemi-framework/docs/modules/modulename.functionname.md

They would be injected as the body in the documentation for the module and the functions in the modules respectively. This way we can customize the documentation for each function in the modules.md file and keep that additional documentation nicely separated.

**Comment by miconda** [comment]
**For point 1)** -- I was not seeing it properly first time, because of the link to module readme, I thought that the prototypes there are the ones from native cfg script, not for kemi.

But it is not one-to-one relation between these functions, some links will point to an invalid anchor. Maybe this can be adjusted later, an invalid anchor should end up at the top of the page.

**For point 2)** -- maybe we can have both variants, with a single file for all to just generate the list of functions, and the variant with one file per module that will allow edits/additions from others. The goal is to have more than just the prototype there, to add description and examples of use.

There should be over 1000 functions once all are exported, that's a lot to have in a single file. Definitely this is not an option for editing it.

I also find it easier to scroll a ToC on the side for module name, even with 200 entries, than scrolling a very long page of 1000-2000 sections.

Proof checking per module should be easier when a new set of split files are added, because we can search by section title regexp in old and new files and compare the results.

In summary:

  * have the variant with one file per module as the option to get more details over the time
  * have the variant with one file for all modules as a reference list -- name and prototype -- eventually pointing inside the one-file-per-module, titles should match in this case

**Comment by verticelo** [comment]
1) Yes, some anchors will lead to the top of the document, such as cases where kemi exports a function which only takes on of the optional params etc.

2) I will make some changes and submit a PR soon which shows my thinking, I don't think I've expressed myself clearly and it feels like we are saying almost the same thing.

**Comment by verticelo** [comment]
Ok, I've added a sample now on customized documentation for:
* The modules.md header
* The RTPEngine module header
* The function set_rtpengine_set2 in the RTPEngine module

Please run your mkdocs --serve and check it out locally on your machine to see how it looks.

This allows anyone to customize the documentation for each exported function and also we will be able to track each change in a separate file.

The process to update the documentation would be:
1) Update the correct modules/module.function.md file and commit it
2) Travis or CircleCi etc runs the pythonscript to generate the module(s) file(s) and uploads the updated docs to the server hosting them

Would this be acceptable?

**Comment by miconda** [comment]
Having a file per function can be also too much, a folder with thousands of small files, that's why I am more for having a file per module, it's matching the style we have for current docs for modules.

Anyhow, I am merging this one and see the results, we can tune afterwards.

Again, thanks for this work!
